Server configuration migration

veggiesaurus opened the pull request:
Moves configuration out of the server source code itself (`config.ts`) to an external `config.json` file (by default, server looks for it in `/etc/carta/config.json`, but this can be changed by using `--config` or `-c` command-line argument.

The schema for the config file is carefully documented, and there's a script to autogenerate HTML documentation from the schema.

ajm-asiaa commented:
@veggiesaurus 
I added the appropriate "google" related entries ("clientID, "validDomains", and "userLookupTable") to the new config.json.
Then I notice this version doesn't seem to have a `public/index.html`. 
Is it being handled a different way now?

ajm-asiaa commented:
@veggiesaurus This version doesn't seem to have `public/index.html`. Is it handled a different way now?

veggiesaurus commented:
> @veggiesaurus This version doesn't seem to have `public/index.html`. Is it handled a different way now?

yes, it now uses `views/index.pug` as a template and auto-generates it depending on your Google client ID :)

ajm-asiaa commented:
Here are a few comments:

1. The `config_schema.html` page is very useful, but I need to clone the repository to my local machine in order to look at it because I can't easily open that page directly on the remote server. Maybe ultimately it could hosted in our online user manual with a link from the README?

2. I'm still not so familiar with JSON. I know `useEmailAsId` already defaults as `true` and it says it is `Type: boolean`. At first I was setting it as `"useEmailAsId": "true",`. Now, after trial and error, I know that Type: boolean must mean that it should be `"useEmailAsId": true,` without the quotes. Maybe `useEmailAsId` could have an Example at least?

3. For `backendPorts`, `min` and `max` are "Required", but that is only apparent after expanding `backendPorts` so "Required" should be added to  `backendPorts`.

4. Tiny bug with `frontendPath`; I see the default is `/etc/carta/frontend`. I didn't have it configured in config.json. The NodeJS log was saying `Serving CARTA frontend from /etc/carta/frontend` but the frontend was still loading even though I had nothing in `/etc/carta/frontend`. I only had the symlink to the frontend in `carta-node-server/public` as we used in the old version. When I removed the symlink, it correctly said `Cannot GET /frontend/`. Well I know this won't be an issue in practice because when the final version is released, users wouldn't know about putting the frontend in `carta-node-server/public/frontend` in the first place.

At the moment I'm still trying to get it to work. I'm getting the connection problem again. I must be missing something in the set up. 

> Proxy error:
> Error: connect ECONNREFUSED 127.0.0.1:3010

ajm-asiaa commented:
Ah I figured it out. I forgot to modify `/etc/sudoers.d/carta_server` for the new location of the carta_backend executable...

I guess this may not be an issue for anyone else as they would probably set "processCommand" and `/etc/sudoers.d/carta_server` with the same path the first time as they work through the instructions.

But still, I wonder if it would it be possible to pick up on this error? It is the second time it has happened to me. Could it be possible that if the NodeJS log reports `errno: 'ECONNREFUSED', code: 'ECONNREFUSED',`, then a suggestion such as 
"Did you configure /etc/sudoers.d/carta_server to be the same as processCommand?" could be printed out?
Or maybe not... I suppose there could be many reasons for this error and it would be impossible to list them all.

ajm-asiaa commented:
@veggiesaurus Here is another small comment. In the previous version, I customised the Dashboard's index.html. One change was the size of the Google Login button. I made it bigger to be more consistent in size with the OPEN CARTA, STOP CARTA SERVER, and LOGOUT buttons.

This is the default appearance:
<img width="338" alt="Screenshot 2020-07-28 at 16 25 27" src="https://user-images.githubusercontent.com/20802551/88639547-7c561980-d0ef-11ea-8ca0-3c39d73de6d1.png">

I made it bigger like this:
<img width="349" alt="Screenshot 2020-07-28 at 16 23 08" src="https://user-images.githubusercontent.com/20802551/88639706-a7d90400-d0ef-11ea-9ece-145f28455258.png">

So it more closely matches the size of the next set of buttons:
<img width="349" alt="Screenshot 2020-07-28 at 16 23 23" src="https://user-images.githubusercontent.com/20802551/88639769-b9baa700-d0ef-11ea-9b0c-8be0ceea5eda.png">

I believe I did so with `data-width` and `data-height`:

>                 <div class="sso-buttons" style="vertical-align:middle; text-align:center">
>                         <div class="g-signin2" data-width="300" data-height="50" data-onsuccess="onSignIn" data-theme="dark"></div>
>                 </div>

I know it is not so important, but it would be a nice visual touch.

veggiesaurus commented:
@ajm-asiaa, I've updated the Google sign in CSS based on your suggestion :+1: 

> 1. The `config_schema.html` page is very useful, but I need to clone the repository to my local machine in order to look at it because I can't easily open that page directly on the remote server. Maybe ultimately it could hosted in our online user manual with a link from the README?

I think that'd be a good idea once it's settled down a bit. I'll leave it for another PR for now

> 2. I'm still not so familiar with JSON. I know `useEmailAsId` already defaults as `true` and it says it is `Type: boolean`. At first I was setting it as `"useEmailAsId": "true",`. Now, after trial and error, I know that Type: boolean must mean that it should be `"useEmailAsId": true,` without the quotes. Maybe `useEmailAsId` could have an Example at least?

I've added `true` and `false` as examples

> 3. For `backendPorts`, `min` and `max` are "Required", but that is only apparent after expanding `backendPorts` so "Required" should be added to  `backendPorts`.

`min` and `max` are required, but `backendPorts` itself is not required, so you can just omit the entire field.

> 4. Tiny bug with `frontendPath`; I see the default is `/etc/carta/frontend`. I didn't have it configured in config.json. The NodeJS log was saying `Serving CARTA frontend from /etc/carta/frontend` but the frontend was still loading even though I had nothing in `/etc/carta/frontend`. I only had the symlink to the frontend in `carta-node-server/public` as we used in the old version. When I removed the symlink, it correctly said `Cannot GET /frontend/`. Well I know this won't be an issue in practice because when the final version is released, users wouldn't know about putting the frontend in `carta-node-server/public/frontend` in the first place.
> 
> At the moment I'm still trying to get it to work. I'm getting the connection problem again. I must be missing something in the set up.
> 
> > Proxy error:
> > Error: connect ECONNREFUSED 127.0.0.1:3010

I think one of the next steps (in a separate PR) is to do some better testing on server startup to ensure that it _can_ actually start the backend correctly, and that the log location is valid and so on

robsimmonds commented:
It works for me. It is much more clear now the config is outside of the ts file. I did mess up the json config on the first try and the messages put out by json the parser can be a little difficult to interpret, but that does not stop this from being a good configuration solution.

confluence commented on README.md:
````diff
@@ -28,10 +32,9 @@ openssl rsa -in carta_private.pem -outform PEM -pubout -out carta_public.pem
 ```
 
 ## Server Configuration
+Server configuration is handled by a configuration file in JSON format, adhering to the [CARTA config schema](config/config_schema.json). Additional details can be found in the auto-generated [config documentation](docs/config_schema.html) or the [example config](config/example_config.json). By default, the server assumes the config file is located at `/etc/carta/config.json`, but this can be cha ...
````
"this can be changed by using" --> "you can change this with"

veggiesaurus replied:
> It would still be good to move to a single script being called by sudoers, but this is good for the configuration update

agreed. That can be done in a separate PR I guess.

veggiesaurus commented:
> It would still be good to move to a single script being called by sudoers, but this is good for the configuration update

agreed. That can be done in a separate PR I guess.
